### PR TITLE
Feat/145/admin announcement dms

### LINKED ISSUE #145: automate sending DMs to drafted quantifiers
It's been proven ineffective to tag quantifiers within discord - it gets lost in the sea of notifications and quantifiers won't realize they have been drafted to quantify praise, which is important since this is a time sensitive process. 

It would be cool if the praise bot could send DMs to quantifiers that have been drafted to quantify in the praise dashboard notifying them they have been drafted and need to visit the dashboard and begin quantification

### PR DESCRIPTION
**Pull request by Vyvy-vi**
Resolves #145

This PR adds quantifier DM functionality to the `discord-bot` package.
(there are some refactors needed here, since there's room for improvement)

https://user-images.githubusercontent.com/62864373/160856201-032576ac-428e-4614-b8bb-96fe319032d0.mov

### CODE REVIEW

**Comment by Vyvy-vi** [comment]
*drafted the PR to fix a timeout bug

**Comment by kristoferlund** [comment]
This also could be simplified and made cleaner:

image

image

I also wonder, what is the reasoning behind having to confirm at all? When sending a regular Discord message you don't get to confirm message before sending.

**Comment by kristoferlund** [comment]
I'd like to propose some text changes.
image

~~Select Target users from dropdown menu~~
Select user group

image

- Capitalisation of terms not needed, except for "Praise"
- Made the texts slightly more direct and shortened to fit into dropdown

**All users**
Send to all activated Praise users

**All quantifiers**
Send to all quantifiers

**All assigned quantifiers**
Send to all quantifiers assigned to a period

**All unfinished quantifiers**
Send to all unfinished quantifiers assigned to a period

image

~~Select Period from dropdown menu~~
Select period

~~Period with end date - xxx~~
End date: xxx

File: packages/discord-bot/src/commands/praiseAdmin.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import { Command } from '../interfaces/Command';
+
+export const praiseAdmin: Command = {
+  data: new SlashCommandBuilder()
+    .setName('admin')
+    .setDescription('Commands to perform admin actions for Praise')
+    .addSubcommand(
+      new SlashCommandSubcommandBuilder()
+        .setName('announce')
+        .setDescription("Automatically announce messages in Users' DMs")
```
Comment (kristoferlund) [inline packages/discord-bot/src/commands/praiseAdmin.ts:16]: "Publish announcements distributed as direct messages to Praise users."

File: packages/discord-bot/src/commands/praiseAdmin.ts [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+        .addStringOption((option) =>
+          option
+            .setName('message')
+            .setDescription(
+              'The message content that you want to send in the DMs'
```
Comment (kristoferlund) [inline packages/discord-bot/src/commands/praiseAdmin.ts:21]: "The message content to publish."

File: packages/discord-bot/src/handlers/announce.ts [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    avatarId: user.avatar,
+    platform: 'DISCORD',
+  } as UserAccount;
+  const userAccount = await UserAccountModel.findOneAndUpdate(
+    { accountId: user.id },
+    ua,
+    { upsert: true, new: true }
+  );
+  if (!userAccount.user) {
+    await interaction.editReply(
```
Comment (kristoferlund) [inline packages/discord-bot/src/handlers/announce.ts:34]: Shouldn't you use `notActivatedError`, already defined, to ensure consistent error messages between the commands?

File: packages/discord-bot/src/handlers/announce.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        case 'confirm': {
+          await interaction.editReply({
+            content: 'Which users do you want to send the message to?',
+            components: [new MessageActionRow().addComponents([dmTargetMenu])],
+          });
+          break;
+        }
+        case 'cancel': {
+          await interaction.editReply({
+            content: 'User Cancelled Interaction.',
```
Comment (kristoferlund) [inline packages/discord-bot/src/handlers/announce.ts:67]: "User cancelled Interaction."

File: packages/discord-bot/src/utils/dmTargets.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        failed.push(`<@!${userId}`);
+        continue;
+      }
+      await discordUser.send(message);
+      successful.push(`<@!${userId}>`);
+    } catch (err) {
+      failed.push(`<@!${userId}>`);
+    }
+  }
+  await interaction.editReply({
```
Comment (kristoferlund) [inline packages/discord-bot/src/utils/dmTargets.ts:39]: Return different message if `successful.length() === 0`. See other comment for idea about message.

File: packages/discord-bot/src/utils/dmTargets.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      const users = await UserModel.find({});
+      await sendDMs(interaction, users, message);
+      return;
+    }
+    case 'QUANTIFIERS': {
+      const users = await UserModel.find({ roles: UserRole.QUANTIFIER });
+      await sendDMs(interaction, users, message);
+      return;
+    }
+    case 'DRAFTED-QUANTIFIERS':
```
Comment (kristoferlund) [inline packages/discord-bot/src/utils/dmTargets.ts:65]: The terms we have been using are:
- Assigned quantifiers
- Unfinished quantifiers

**Review by kristoferlund — CHANGES_REQUESTED** [review]
Great work @Vyvy-vi !!

Just needs some finishing touches - see proposed text changes etc.

**Comment by kristoferlund** [comment]
Moving the confirmation to the last step, how about this @Vyvy-vi ?

image

File: packages/discord-bot/src/commands/praiseAdmin.ts (hunk 2 above)
Comment (Vyvy-vi) [reply packages/discord-bot/src/commands/praiseAdmin.ts:21]: how about "The message content to announce."?

**Review by kristoferlund — APPROVED** [review]
Great work!! 🎉

**Pull request merged**